File: glean-core/python/glean/glean.py
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

"""
The main Glean general API.
"""


import atexit
import json
import logging
from pathlib import Path
import platform
import shutil
import tempfile
import threading
from typing import Dict, List, Optional, Set, TYPE_CHECKING


from .config import Configuration
from ._dispatcher import Dispatcher
from . import _ffi
from .net import PingUploadWorker
from ._process_dispatcher import ProcessDispatcher
from . import _util


# To avoid cyclical imports, but still make mypy type-checking work.
# See https://mypy.readthedocs.io/en/latest/common_issues.html#import-cycles
if TYPE_CHECKING:
    from .metrics import PingType, RecordedExperimentData


log = logging.getLogger(__name__)


def _rmtree(path) -> bool:
    """
    A small wrapper around shutil.rmtree to make it runnable on the
    ProcessDispatcher.
    """
    shutil.rmtree(path)
    return True


class Glean:
    """
    The main Glean API.

    Before any data collection can take place, the Glean SDK **must** be
    initialized from the application.

    >>> Glean.initialize(
    ...     application_id="my-app",
    ...     application_version="0.0.0",
    ...     upload_enabled=True
    ... )
    """

    # Whether Glean was initialized
    _initialized = False  # type: bool

    # The Configuration that was passed to `initialize`
    _configuration = None  # type: Configuration

    # The directory that Glean stores data in
    _data_dir = Path()  # type: Path

    # Whether Glean "owns" the data directory and should destroy it upon reset.
    _destroy_data_dir = False  # type: bool

    # Keep track of this setting before Glean is initialized
    _upload_enabled = True  # type: bool

    # The ping types, so they can be registered prior to Glean initialization,
    # and saved between test runs.
    _ping_type_queue = set()  # type: Set[PingType]

    # The application id to send in the ping.
    _application_id = None  # type: str

    # The version of the application sending Glean data.
    _application_version = None  # type: str

    # A thread lock for Glean operations that need to be synchronized
    _thread_lock = threading.RLock()

    @classmethod
    def initialize(
        cls,
        application_id: str,
        application_version: str,
        upload_enabled: bool,
        configuration: Optional[Configuration] = None,
        data_dir: Optional[Path] = None,
    ) -> None:
        """
        Initialize the Glean SDK.

        This should only be initialized once by the application, and not by
        libraries using the Glean SDK. A message is logged to error and no
        changes are made to the state if initialize is called a more than
        once.

        Args:
            application_id (str): The application id to use when sending pings.
            application_version (str): The version of the application sending
                Glean data.
            upload_enabled (bool): Controls whether telemetry is enabled. If
                disabled, all persisted metrics, events and queued pings
                (except first_run_date) are cleared.
            configuration (glean.config.Configuration): (optional) An object with
                global settings.
            data_dir (pathlib.Path): (optional) The path to the Glean data
                directory. If not provided, uses a temporary directory.
        """
        with cls._thread_lock:
            if cls.is_initialized():
                return

            atexit.register(Glean._reset)

            if configuration is None:
                configuration = Configuration()

            if data_dir is None:
                data_dir = Path(tempfile.TemporaryDirectory().name)
                cls._destroy_data_dir = True
            else:
                cls._destroy_data_dir = False
            cls._data_dir = data_dir

            cls._configuration = configuration
            cls._application_id = application_id
            cls._application_version = application_version

            cls.set_upload_enabled(upload_enabled)

        # Use `Glean._execute_task` rather than `Glean.launch` here, since we
        # never want to put this work on the `Dispatcher._preinit_queue`.
        @Dispatcher._execute_task
        def initialize():
            # Other platforms register the built-in pings here. That is not
            # necessary on Python since it doesn't have the problem with static
            # initializers that Kotlin and Swift have.

            cfg = _ffi.make_config(
                cls._data_dir,
                application_id,
                cls._upload_enabled,
                configuration.max_events,
            )

            cls._initialized = _ffi.lib.glean_initialize(cfg) != 0

            # If initialization of Glean fails, we bail out and don't initialize
            # further
            if not cls._initialized:
                return

            # Kotlin bindings have a "synchronized" here, but that is
            # unnecessary given that Python has a GIL.
            with cls._thread_lock:
                for ping in cls._ping_type_queue:
                    cls.register_ping_type(ping)

            # If this is the first time ever the Glean SDK runs, make sure to set
            # some initial core metrics in case we need to generate early pings.
            # The next times we start, we would have them around already.
            is_first_run = _ffi.lib.glean_is_first_run() != 0
            if is_first_run:
                cls._initialize_core_metrics()

            # Deal with any pending events so we can start recording new ones
            if (
                _ffi.lib.glean_on_ready_to_submit_pings()
                or cls._upload_enabled is False
            ):
                PingUploadWorker.process()

            # Glean Android sets up the metrics ping scheduler here, but we don't
            # have one.

            # Other platforms check for the "dirty bit" and send the `baseline` ping
            # with reason `dirty_startup`.

            # From the second time we run, after all startup pings are generated,
            # make sure to clear `lifetime: application` metrics and set them again.
            # Any new value will be sent in newly generated pings after startup.
            if not is_first_run:
                _ffi.lib.glean_clear_application_lifetime_metrics()
                cls._initialize_core_metrics()

            Dispatcher.flush_queued_initial_tasks()

            # Glean Android sets up the lifecycle observer here. We don't really
            # have a lifecycle.

    @_util.classproperty
    def configuration(cls) -> Configuration:
        """
        Access the configuration object to change dynamic parameters.
        """
        return cls._configuration

    @classmethod
    def _reset(cls) -> None:
        """
        Resets the Glean singleton.
        """
        # TODO: 1594184 Send the metrics ping

        # WARNING: Do not run any tasks on the Dispatcher from here since this
        # is called atexit.

        # Wait for the dispatcher thread to complete.
        Dispatcher._task_worker._shutdown_thread()

        Dispatcher.reset()

        # Wait for the subprocess to complete.  We only need to do this if
        # we know we are going to be deleting the data directory.
        if cls._destroy_data_dir and cls._data_dir.exists():
            ProcessDispatcher._wait_for_last_process()

        # Destroy the Glean object.
        # Importantly on Windows, this closes the handle to the database so
        # that the data directory can be deleted without a multiple access
        # violation.
        if cls._initialized:
            _ffi.lib.glean_destroy_glean()
        cls._initialized = False

        # Remove the atexit handler or it will get called multiple times at
        # exit.
        atexit.unregister(cls._reset)

        if cls._destroy_data_dir and cls._data_dir.exists():
            # This needs to be run in the same one-at-a-time process as the
            # PingUploadWorker to avoid a race condition. This will block the
            # main thread waiting for all pending uploads to complete, but this
            # only happens during testing when the data directory is a
            # temporary directory, so there is no concern about delaying
            # application shutdown here.
            p = ProcessDispatcher.dispatch(_rmtree, (str(cls._data_dir),))
            p.wait()

    @classmethod
    def is_initialized(cls) -> bool:
        """
        Returns True if the Glean SDK has been initialized.
        """
        return cls._initialized

    @classmethod
    def register_ping_type(cls, ping: "PingType") -> None:
        """
        Register the ping type in the registry.
        """
        with cls._thread_lock:
            if cls.is_initialized():
                _ffi.lib.glean_register_ping_type(ping._handle)

            # We need to keep track of pings, so they get re-registered after a
            # reset. This state is kept across Glean resets, which should only
            # ever happen in test mode. It's a set and keeping them around
            # forever should not have much of an impact.
            cls._ping_type_queue.add(ping)

    @classmethod
    def test_has_ping_type(cls, ping_name: str) -> bool:
        """
        Returns True if a ping by this name is in the ping registry.
        """
        return bool(
            _ffi.lib.glean_test_has_ping_type(_ffi.ffi_encode_string(ping_name))
        )

    @classmethod
    def set_upload_enabled(cls, enabled: bool) -> None:
        """
        Enable or disable Glean collection and upload.

        Metric collection is enabled by default.

        When uploading is disabled, metrics aren't recorded at all and no data
        is uploaded.

        When disabling, all pending metrics, events and queued pings are cleared.

        When enabling, the core Glean metrics are recreated.

        Args:
            enabled (bool): When True, enable metric collection.
        """
        if cls.is_initialized():
            original_enabled = cls.get_upload_enabled()

            @Dispatcher.launch
            def set_upload_enabled():
                _ffi.lib.glean_set_upload_enabled(enabled)

                if original_enabled is False and cls.get_upload_enabled() is True:
                    cls._initialize_core_metrics()

                if original_enabled is True and cls.get_upload_enabled() is False:
                    # If uploading is disabled, we need to send the deletion-request ping
                    PingUploadWorker.process()

        else:
            cls._upload_enabled = enabled

    @classmethod
    def get_upload_enabled(cls) -> bool:
        """
        Get whether or not Glean is allowed to record and upload data.
        """
        if cls.is_initialized():
            return bool(_ffi.lib.glean_is_upload_enabled())
        else:
            return cls._upload_enabled

    @classmethod
    def set_experiment_active(
        cls, experiment_id: str, branch: str, extra: Optional[Dict[str, str]] = None
    ) -> None:
        """
        Indicate that an experiment is running. Glean will then add an
        experiment annotation to the environment which is sent with pings. This
        information is not persisted between runs.

        Args:
            experiment_id (str): The id of the active experiment (maximum 100
                bytes)
            branch (str): The experiment branch (maximum 100 bytes)
            extra (dict of str -> str): Optional metadata to output with the
                ping
        """
        if extra is None:
            keys = []  # type: List[str]
            values = []  # type: List[str]
        else:
            keys, values = zip(*extra.items())  # type: ignore

        @Dispatcher.launch
        def set_experiment_active():
            _ffi.lib.glean_set_experiment_active(
                _ffi.ffi_encode_string(experiment_id),
                _ffi.ffi_encode_string(branch),
                _ffi.ffi_encode_vec_string(keys),
                _ffi.ffi_encode_vec_string(values),
                len(keys),
            )

    @classmethod
    def set_experiment_inactive(cls, experiment_id: str) -> None:
        """
        Indicate that the experiment is no longer running.

        Args:
            experiment_id (str): The id of the experiment to deactivate.
        """

        @Dispatcher.launch
        def set_experiment_inactive():
            _ffi.lib.glean_set_experiment_inactive(
                _ffi.ffi_encode_string(experiment_id)
            )

    @classmethod
    def test_is_experiment_active(cls, experiment_id: str) -> bool:
        """
        Tests whether an experiment is active, for testing purposes only.

        Args:
            experiment_id (str): The id of the experiment to look for.

        Returns:
            is_active (bool): If the experiement is active and reported in
                pings.
        """
        return bool(
            _ffi.lib.glean_experiment_test_is_active(
                _ffi.ffi_encode_string(experiment_id)
            )
        )

    @classmethod
    def test_get_experiment_data(cls, experiment_id: str) -> "RecordedExperimentData":
        """
        Returns the stored data for the requested active experiment, for testing purposes only.

        Args:
            experiment_id (str): The id of the experiment to look for.

        Returns:
            experiment_data (RecordedExperimentData): The data associated with
                the experiment.
        """
        from .metrics import RecordedExperimentData

        json_string = _ffi.ffi_decode_string(
            _ffi.lib.glean_experiment_test_get_data(
                _ffi.ffi_encode_string(experiment_id)
            )
        )

        json_tree = json.loads(json_string)

        return RecordedExperimentData(**json_tree)  # type: ignore

    @classmethod
    def _initialize_core_metrics(cls) -> None:
        """
        Set a few metrics that will be sent as part of every ping.
        """
        from ._builtins import metrics

        metrics.glean.baseline.locale.set(_util.get_locale_tag())
        metrics.glean.internal.metrics.os_version.set(platform.release())
        metrics.glean.internal.metrics.architecture.set(platform.machine())
        metrics.glean.internal.metrics.locale.set(_util.get_locale_tag())

        if cls._configuration.channel is not None:
            metrics.glean.internal.metrics.app_channel.set(cls._configuration.channel)

        metrics.glean.internal.metrics.app_build.set(cls._application_id)

        if cls._application_version is not None:
            metrics.glean.internal.metrics.app_display_version.set(
                cls._application_version
            )

    @classmethod
    def get_data_dir(cls) -> Path:
        """
        Get the data directory for Glean.
        """
        return cls._data_dir

    @classmethod
    def test_collect(cls, ping: "PingType", reason: Optional[str] = None) -> str:
        """
        Collect a ping and return as a string.

        Args:
            ping: The PingType to submit
            reason (str, optional): The reason code to record in the ping.
        """
        return _ffi.ffi_decode_string(
            _ffi.lib.glean_ping_collect(ping._handle, _ffi.ffi_encode_string(reason))
        )

    @classmethod
    def _submit_ping(cls, ping: "PingType", reason: Optional[str] = None) -> None:
        """
        Collect and submit a ping for eventual uploading.

        If the ping currently contains no content, it will not be assembled and
        queued for sending.

        Args:
            ping (PingType): Ping to submit.
            reason (str, optional): The reason the ping was submitted.
        """
        cls._submit_ping_by_name(ping.name, reason)

    @classmethod
    @Dispatcher.task
    def _submit_ping_by_name(cls, ping_name: str, reason: Optional[str] = None) -> None:
        """
        Collect and submit a ping by name for eventual uploading.

        The ping will be looked up in the known instances of
        `glean.metrics.PingType`. If the ping isn't known, an error is logged
        and the ping isn't queued for uploading.

        If the ping currently contains no content, it will not be assembled and
        queued for sending.

        Args:
            ping_name (str): Ping name to submit.
            reason (str, optional): The reason code to include in the ping.
        """
        if not cls.is_initialized():
            log.error("Glean must be initialized before submitting pings.")
            return

        if not cls.get_upload_enabled():
            log.error("Glean disabled: not submitting any pings.")
            return

        sent_ping = _ffi.lib.glean_submit_ping_by_name(
            _ffi.ffi_encode_string(ping_name), _ffi.ffi_encode_string(reason),
        )

        if sent_ping:
            PingUploadWorker.process()


__all__ = ["Glean"]

File: glean-core/csharp/Glean/Glean.cs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using Mozilla.Glean.FFI;
using Mozilla.Glean.Net;
using Mozilla.Glean.Private;
using Serilog;
using static Mozilla.Glean.GleanMetrics.GleanInternalMetricsOuter;
using static Mozilla.Glean.GleanPings.GleanInternalPingsOuter;
using static Mozilla.Glean.Utils.GleanLogger;

namespace Mozilla.Glean
{
    /// <summary>
    /// The Glean Gneral API.
    /// </summary>
    public sealed class Glean
    {
        // Initialize the singleton using the `Lazy` facilities.
        private static readonly Lazy<Glean>
          lazy = new Lazy<Glean>(() => new Glean());
        public static Glean GleanInstance => lazy.Value;

        private bool initialized = false;

        // Keep track of ping types that have been registered before Glean is initialized.
        private HashSet<PingTypeBase> pingTypeQueue = new HashSet<PingTypeBase>();

        private Configuration configuration;

        // This is the wrapped http uploading mechanism: provides base functionalities
        // for logging and delegates the actual upload to the implementation in
        // the `Configuration`.
        private BaseUploader httpClient;

        // The version of the application sending Glean data.
        private string applicationVersion;

        /// <summary>
        /// This is the tag used for logging from this class.
        /// </summary>
        private const string LogTag = "glean/Glean";

        /// <summary>
        /// This is the name of the language used by this Glean binding.
        /// </summary>
        private readonly static string LanguageBindingName = "C#";

        /// <summary>
        /// A logger configured for this class
        /// </summary>
        private static readonly ILogger Log = GetLogger(LogTag);

        private Glean()
        {
            // Private constructor to disallow instantiation since
            // this is meant to be a singleton. It only wires up the
            // glean-core logging on the Rust side.
            LibGleanFFI.glean_enable_logging();
        }

        /// <summary>
        /// Initialize the Glean SDK.
        ///
        /// This should only be initialized once by the application, and not by
        /// libraries using the Glean SDK. A message is logged to error and no
        /// changes are made to the state if initialize is called a more than
        /// once.
        ///
        /// This method must be called from the main thread.
        /// </summary>
        /// <param name="applicationId">The application id to use when sending pings.</param>
        /// <param name="applicationVersion">The version of the application sending
        /// Glean data.</param>
        /// <param name="uploadEnabled">A `bool` that determines whether telemetry is enabled.
        /// If disabled, all persisted metrics, events and queued pings (except first_run_date)
        /// are cleared.</param>
        /// <param name="configuration">A Glean `Configuration` object with global settings</param>
        /// <param name="dataDir">The path to the Glean data directory.</param>
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void Initialize(
            string applicationId,
            string applicationVersion,
            bool uploadEnabled,
            Configuration configuration,
            string dataDir
            )
        {
            /*
            // Glean initialization must be called on the main thread, or lifecycle
            // registration may fail. This is also enforced at build time by the
            // @MainThread decorator, but this run time check is also performed to
            // be extra certain.
            ThreadUtils.assertOnUiThread()

            // In certain situations Glean.initialize may be called from a process other than the main
            // process.  In this case we want initialize to be a no-op and just return.
            if (!isMainProcess(applicationContext)) {
                Log.e(LOG_TAG, "Attempted to initialize Glean on a process other than the main process")
                return
            }

            this.applicationContext = applicationContext*/

            if (IsInitialized()) {
                Log.Error("Glean should not be initialized multiple times");
                return;
            }

            this.configuration = configuration;
            this.applicationVersion = applicationVersion;
            httpClient = new BaseUploader(configuration.httpClient);
            // this.gleanDataDir = File(applicationContext.applicationInfo.dataDir, GLEAN_DATA_DIR)

            Dispatchers.ExecuteTask(() =>
            {
                RegisterPings(GleanInternalPings);

                IntPtr maxEventsPtr = IntPtr.Zero;
                if (configuration.maxEvents != null)
                {
                    maxEventsPtr = Marshal.AllocHGlobal(sizeof(int));
                    // It's safe to call `configuration.maxEvents.Value` because we know
                    // `configuration.maxEvents` is not null.
                    Marshal.WriteInt32(maxEventsPtr, configuration.maxEvents.Value);
                }

                LibGleanFFI.FfiConfiguration cfg = new LibGleanFFI.FfiConfiguration
                {
                    data_dir = dataDir,
                    package_name = applicationId,
                    language_binding_name = LanguageBindingName,
                    upload_enabled = uploadEnabled,
                    max_events = maxEventsPtr,
                    delay_ping_lifetime_io = false
                };

                // To work around a bug in the version of Mono shipped with Unity 2019.4.1f1,
                // copy the FFI configuration structure to unmanaged memory and pass that over
                // to glean-core, otherwise calling `glean_initialize` will crash and have
                // `__icall_wrapper_mono_struct_delete_old` in the stack. See bug 1648784 for
                // more details.
                IntPtr ptrCfg = Marshal.AllocHGlobal(Marshal.SizeOf(cfg));
                Marshal.StructureToPtr(cfg, ptrCfg, false);

                initialized = LibGleanFFI.glean_initialize(ptrCfg) != 0;

                // This is safe to call even if `maxEventsPtr = IntPtr.Zero`.
                Marshal.FreeHGlobal(maxEventsPtr);
                // We were able to call `glean_initialize`, free the memory allocated for the
                // FFI configuration object.
                Marshal.FreeHGlobal(ptrCfg);

                // If initialization of Glean fails we bail out and don't initialize further.
                if (!initialized)
                {
                    return;
                }

                // If any pings were registered before initializing, do so now.
                // We're not clearing this queue in case Glean is reset by tests.
                lock (this)
                {
                    foreach (var ping in pingTypeQueue)
                    {
                        RegisterPingType(ping);
                    }
                }

                // If this is the first time ever the Glean SDK runs, make sure to set
                // some initial core metrics in case we need to generate early pings.
                // The next times we start, we would have them around already.
                bool isFirstRun = LibGleanFFI.glean_is_first_run() != 0;
                if (isFirstRun)
                {
                    InitializeCoreMetrics();
                }


                // Deal with any pending events so we can start recording new ones
                bool pingSubmitted = LibGleanFFI.glean_on_ready_to_submit_pings() != 0;

                // We need to enqueue the BaseUploader in these cases:
                // 1. Pings were submitted through Glean and it is ready to upload those pings;
                // 2. Upload is disabled, to upload a possible deletion-request ping.
                if (pingSubmitted || !uploadEnabled)
                {
                    httpClient.TriggerUpload(configuration);
                }
                /*
                // Set up information and scheduling for Glean owned pings. Ideally, the "metrics"
                // ping startup check should be performed before any other ping, since it relies
                // on being dispatched to the API context before any other metric.
                metricsPingScheduler = MetricsPingScheduler(applicationContext)
                metricsPingScheduler.schedule()

                // Check if the "dirty flag" is set. That means the product was probably
                // force-closed. If that's the case, submit a 'baseline' ping with the
                // reason "dirty_startup". We only do that from the second run.
                if (!isFirstRun && LibGleanFFI.INSTANCE.glean_is_dirty_flag_set().toBoolean()) {
                    submitPingByNameSync("baseline", "dirty_startup")
                    // Note: while in theory we should set the "dirty flag" to true
                    // here, in practice it's not needed: if it hits this branch, it
                    // means the value was `true` and nothing needs to be done.
                }*/

                // From the second time we run, after all startup pings are generated,
                // make sure to clear `lifetime: application` metrics and set them again.
                // Any new value will be sent in newly generated pings after startup.
                if (!isFirstRun)
                {
                    LibGleanFFI.glean_clear_application_lifetime_metrics();
                    InitializeCoreMetrics();
                }

                // Signal Dispatcher that init is complete
                Dispatchers.FlushQueuedInitialTasks();
                /*
                // At this point, all metrics and events can be recorded.
                // This should only be called from the main thread. This is enforced by
                // the @MainThread decorator and the `assertOnUiThread` call.
                MainScope().launch {
                    ProcessLifecycleOwner.get().lifecycle.addObserver(gleanLifecycleObserver)
                }*/
            });
        }

        /// <summary>
        /// Whether or not the Glean SDK has been initialized.
        /// </summary>
        /// <returns>Returns true if the Glean SDK has been initialized.</returns>
        internal bool IsInitialized()
        {
            return initialized;
        }

        /// <summary>
        /// Enable or disable Glean collection and upload.
        ///
        /// Metric collection is enabled by default.
        ///
        /// When uploading is disabled, metrics aren't recorded at all and no data
        /// is uploaded.
        ///
        /// When disabling, all pending metrics, events and queued pings are cleared
        /// and a `deletion-request` is generated.
        ///
        /// When enabling, the core Glean metrics are recreated.
        /// </summary>
        /// <param name="enabled">When `true`, enable metric collection.</param>
        public void SetUploadEnabled(bool enabled)
        {
            // Changing upload enabled always happens asynchronous.
            // That way it follows what a user expect when calling it inbetween other calls:
            // It executes in the right order.
            //
            // Because the dispatch queue is halted until Glean is fully initialized
            // we can safely enqueue here and it will execute after initialization.
            Dispatchers.LaunchAPI(() => {
                bool originalEnabled = this.GetUploadEnabled();
                LibGleanFFI.glean_set_upload_enabled(enabled);

                if (!enabled)
                {
                    // Cancel any pending workers here so that we don't accidentally upload or
                    // collect data after the upload has been disabled.
                    // TODO: metricsPingScheduler.cancel()

                    // Cancel any pending workers here so that we don't accidentally upload
                    // data after the upload has been disabled.
                    httpClient.CancelUploads();
                }

                if (!originalEnabled && enabled)
                {
                    // If uploading is being re-enabled, we have to restore the
                    // application-lifetime metrics.
                    InitializeCoreMetrics();
                }

                if (originalEnabled && !enabled)
                {
                    // If uploading is disabled, we need to send the deletion-request ping
                    httpClient.TriggerUpload(configuration);
                }
            });
        }

        /// <summary>
        /// Get whether or not Glean is allowed to record and upload data.
        ///
        /// Caution: the result is only correct if Glean is already initialized.
        /// </summary>
        /// <returns>`true` if Glean is allowed to record and upload data.</returns>
        public bool GetUploadEnabled()
        {
            if (IsInitialized())
            {
                return LibGleanFFI.glean_is_upload_enabled() != 0;
            }
            else
            {
                return true;
            }
        }

        /// <summary>
        /// TEST ONLY FUNCTION.
        /// Resets the Glean state and triggers init again.
        /// </summary>
        /// <param name="applicationId">The application id to use when sending pings.</param>
        /// <param name="applicationVersion">The version of the application sending
        /// Glean data.</param>
        /// <param name="uploadEnabled">A `bool` that determines whether telemetry is enabled.
        /// If disabled, all persisted metrics, events and queued pings (except first_run_date)
        /// are cleared.</param>
        /// <param name="configuration">A Glean `Configuration` object with global settings</param>
        /// <param name="dataDir">The path to the Glean data directory.</param>
        /// <param name="clearStores">If `true` clear the contents of all stores.</param>
        internal void Reset(
            string applicationId,
            string applicationVersion,
            bool uploadEnabled,
            Configuration configuration,
            string dataDir,
            bool clearStores = true)
        {
            Dispatchers.TestingMode = true;

            if (IsInitialized() && clearStores)
            {
                // Clear all the stored data.
                LibGleanFFI.glean_test_clear_all_stores();
            }

            // TODO: isMainProcess = null

            // Init Glean.
            TestDestroyGleanHandle();
            Initialize(applicationId, applicationVersion, uploadEnabled, configuration, dataDir);
        }

        /// <summary>
        /// TEST ONLY FUNCTION.
        /// Destroy the owned glean-core handle.
        /// </summary>
        internal void TestDestroyGleanHandle()
        {
            if (!IsInitialized())
            {
                // We don't need to destroy Glean: it wasn't initialized.
                return;
            }

            LibGleanFFI.glean_destroy_glean();
            initialized = false;
        }

        private void InitializeCoreMetrics()
        {
            // The `Environment.OSVersion` object will return a version that represents an approximate
            // version of the OS. As the MSDN docs state, this is unreliable:
            // https://docs.microsoft.com/en-us/dotnet/api/system.environment.osversion?redirectedfrom=MSDN&view=netstandard-2.0
            // However, there's really nothing we could do about it. Unless the product using the
            // Glean SDK correctly marks their manifest as Windows 10 compatible, this API will report
            // Windows 8 version (6.2). See the remarks section here:
            // https://docs.microsoft.com/en-gb/windows/win32/api/winnt/ns-winnt-osversioninfoexa?redirectedfrom=MSDN#remarks
            try
            {
                GleanInternalMetrics.osVersion.SetSync(Environment.OSVersion.Version.ToString());
            }
            catch (InvalidOperationException)
            {
                GleanInternalMetrics.osVersion.SetSync("Unknown");
            }

            // Possible values for `RuntimeInformation.OSArchitecture` are documented here:
            // https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.architecture?view=netstandard-2.0
            GleanInternalMetrics.architecture.SetSync(RuntimeInformation.OSArchitecture.ToString());

            try
            {
                // CurrentUiCulture is used for representing the locale of the UI, coming from the OS,
                // while CurrentCulture is the general locale used for other things (e.g. currency).
                GleanInternalMetrics.locale.SetSync(CultureInfo.CurrentUICulture.Name);
            }
            catch (Exception)
            {
                GleanInternalMetrics.locale.SetSync("und");
            }

            if (configuration.channel != null)
            {
                GleanInternalMetrics.appChannel.SetSync(configuration.channel);
            }

            GleanInternalMetrics.appDisplayVersion.SetSync(applicationVersion);
            GleanInternalMetrics.appBuild.SetSync(configuration.buildId ?? "Unknown");
        }

        /// <summary>
        /// Register the pings generated from `ManualPings` with the Glean SDK.
        /// </summary>
        /// <param name="pings"> The `Pings` object generated for your library or application
        /// by the Glean SDK.</param>
        private void RegisterPings(object pings)
        {
            // Instantiating the Pings object to send this function is enough to
            // call the constructor and have it registered through [Glean.registerPingType].
            Log.Information("Registering pings");
        }

        /// <summary>
        /// Handle the background event and send the appropriate pings.
        /// </summary>
        internal void HandleBackgroundEvent()
        {
            GleanInternalPings.baseline.Submit(BaselineReasonCodes.background);
            GleanInternalPings.events.Submit(EventsReasonCodes.background);
        }

        /// <summary>
        /// Collect and submit a ping for eventual upload.
        ///
        /// The ping content is assembled as soon as possible, but upload is not
        /// guaranteed to happen immediately, as that depends on the upload
        /// policies.
        ///
        /// If the ping currently contains no content, it will not be assembled and
        /// queued for sending.
        /// </summary>
        /// <param name="ping">Ping to submit.</param>
        /// <param name="reason">The reason the ping is being submitted.</param>
        internal void SubmitPing(PingTypeBase ping, string reason = null)
        {
            SubmitPingByName(ping.name, reason);
        }

        /// <summary>
        /// Collect and submit a ping for eventual upload by name.
        ///
        /// The ping will be looked up in the known instances of `PingType`. If the
        /// ping isn't known, an error is logged and the ping isn't queued for uploading.
        ///
        /// The ping content is assembled as soon as possible, but upload is not
        /// guaranteed to happen immediately, as that depends on the upload
        /// policies.
        ///
        /// If the ping currently contains no content, it will not be assembled and
        /// queued for sending, unless explicitly specified otherwise in the registry
        /// file.
        /// </summary>
        /// <param name="name">Name of the ping to submit.</param>
        /// <param name="reason">The reason the ping is being submitted.</param>
        internal void SubmitPingByName(string name, string reason = null)
        {
            Dispatchers.LaunchAPI(() =>
            {
                SubmitPingByNameSync(name, reason);
            });
        }

        /// <summary>
        /// Collect and submit a ping (by its name) for eventual upload, synchronously.
        ///
        /// The ping will be looked up in the known instances of `PingType`. If the
        /// ping isn't known, an error is logged and the ping isn't queued for uploading.
        ///
        /// The ping content is assembled as soon as possible, but upload is not
        /// guaranteed to happen immediately, as that depends on the upload
        /// policies.
        ///
        /// If the ping currently contains no content, it will not be assembled and
        /// queued for sending, unless explicitly specified otherwise in the registry
        /// file.
        /// </summary>
        /// <param name="name">Name of the ping to submit.</param>
        /// <param name="reason">The reason the ping is being submitted.</param>
        internal void SubmitPingByNameSync(string name, string reason = null)
        {
            if (!IsInitialized())
            {
                Log.Error("Glean must be initialized before submitting pings.");
                return;
            }

            if (!GetUploadEnabled())
            {
                Log.Error("Glean disabled: not submitting any pings.");
                return;
            }

            bool hasSubmittedPing = Convert.ToBoolean(LibGleanFFI.glean_submit_ping_by_name(name, reason));
            if (hasSubmittedPing)
            {
                httpClient.TriggerUpload(configuration);
            }
        }

        /// <summary>
        /// Register a [PingType] in the registry associated with this [Glean] object.
        /// </summary>
        [MethodImpl(MethodImplOptions.Synchronized)]
        internal void RegisterPingType(PingTypeBase pingType)
        {
            if (IsInitialized())
            {
                LibGleanFFI.glean_register_ping_type(
                    pingType.handle
                );
            }

            // We need to keep track of pings, so they get re-registered after a reset.
            // This state is kept across Glean resets, which should only ever happen in test mode.
            // Or by the instrumentation tests (`connectedAndroidTest`), which relaunches the application activity,
            // but not the whole process, meaning globals, such as the ping types, still exist from the old run.
            // It's a set and keeping them around forever should not have much of an impact.

            pingTypeQueue.Add(pingType);
        }

        /// <summary>
        /// TEST ONLY FUNCTION.
        /// Returns true if a ping by this name is in the ping registry.
        /// </summary>
        internal bool TestHasPingType(string pingName)
        {
            return LibGleanFFI.glean_test_has_ping_type(pingName) != 0;
        }
}
}
